### PR TITLE
Fix too large offsets when point labels are offset from symbol bounds and marker symbol is a non-square symbol

### PR DESCRIPTION
**Pull request by nyalldawson**
Before -- note how the labels to the top right have a much greater distance from the marker vs the label which is placed over the feature:

image

After:
image

### CODE REVIEW

File: src/core/symbology/qgssymbollayer.h [hunk 1]
```diff
@@ -859,6 +860,14 @@ class CORE_EXPORT QgsMarkerSymbolLayer : public QgsSymbolLayer
      */
     virtual QRectF bounds( QPointF point, QgsSymbolRenderContext &context ) = 0;
 
+    /**
+     * Returns the symbol bounds of the marker symbol layer, taking into account
+     * any data defined overrides and offsets which are set for the marker layer.
+     * \returns symbol bounds, in painter units
+     * \since QGIS 3.20
+     */
+    virtual QgsMarkerSymbolBounds symbolBounds( QPointF point, QgsSymbolRenderContext &context );
```
Comment (pblottiere) [inline src/core/symbology/qgssymbollayer.h:869]: Why not `const QPointF &point` in this case?

Comment (nyalldawson) [reply src/core/symbology/qgssymbollayer.h:869]: Qpointf is a small and trivially copiable class -- according to the clazy checks built into qtcreator it's quicker to copy than dereference!

File: src/core/labeling/qgspallabeling.h [hunk 2]
```diff
@@ -827,18 +828,29 @@ class CORE_EXPORT QgsPalLayerSettings
...
      */
-    void registerFeature( const QgsFeature &f, QgsRenderContext &context,
-                          QgsLabelFeature **labelFeature SIP_PYARGREMOVE = nullptr,
-                          QgsGeometry obstacleGeometry SIP_PYARGREMOVE = QgsGeometry(), const QgsSymbol *symbol SIP_PYARGREMOVE = nullptr );
+    void registerFeature( const QgsFeature &f, QgsRenderContext &context );
+
+#ifndef SIP_RUN
+
+    /**
+     * Register a feature for labeling.
```
Comment (pblottiere) [inline src/core/labeling/qgspallabeling.h:840]:
```suggestion
     * Registers a feature for labeling.
```

File: src/core/labeling/qgspallabeling.h [hunk 3]
```diff
@@ -827,18 +828,29 @@ class CORE_EXPORT QgsPalLayerSettings
...
+     * \param feature feature to label
+     * \param context render context. The QgsExpressionContext contained within the render context
+     * must have already had the feature and fields sets prior to calling this method.
```
Comment (pblottiere) [inline src/core/labeling/qgspallabeling.h:843]:
```suggestion
     * must have already had the feature and fields set prior to calling this method.
```

Comment (pblottiere) [reply src/core/labeling/qgspallabeling.h:843]: (i'm not sure)

File: src/core/pal/feature.cpp [hunk 4]
```diff
@@ -430,18 +430,32 @@ std::unique_ptr<LabelPosition> FeaturePart::createCandidatePointOnSurface( Point
...
 }
 
-void createCandidateAtOrderedPositionOverPoint( double &labelX, double &labelY, LabelPosition::Quadrant &quadrant, double x, double y, double labelWidth, double labelHeight, QgsPalLayerSettings::PredefinedPointPosition position, double distanceToLabel, const QgsMargins &visualMargin, double symbolWidthOffset, double symbolHeightOffset )
+void createCandidateAtOrderedPositionOverPoint( double &labelX, double &labelY, LabelPosition::Quadrant &quadrant, double x, double y, double labelWidth, double labelHeight, QgsPalLayerSettings::PredefinedPointPosition position, double distanceToLabel, const QgsMargins &visualMargin, const QgsMarkerSymbolBounds *symbolBounds )
 {
   double alpha = 0.0;
   double deltaX = 0;
   double deltaY = 0;
+
+  double symbolWidthOffset = ( symbolBounds ? symbolBounds->boundingBox().width() / 2.0 : 0.0 );
```
Comment (pblottiere) [inline src/core/pal/feature.cpp:439]:
```suggestion
  const double symbolWidthOffset = ( symbolBounds ? symbolBounds->boundingBox().width() / 2.0 : 0.0 );
```

File: src/core/pal/feature.cpp [hunk 5]
```diff
@@ -430,18 +430,32 @@ std::unique_ptr<LabelPosition> FeaturePart::createCandidatePointOnSurface( Point
...
+  double symbolHeightOffset = ( symbolBounds ? symbolBounds->boundingBox().height() / 2.0 : 0.0 );
```
Comment (pblottiere) [inline src/core/pal/feature.cpp:440]:
```suggestion
  const double symbolHeightOffset = ( symbolBounds ? symbolBounds->boundingBox().height() / 2.0 : 0.0 );
```

**Comment by pblottiere** [comment]
Code style looks good to me :+1:

**Pull request closed without merging**